gym_roboy/envs/ros_proxy.py

```diff
@@ -68,6 +70,23 @@ def __init__(self, timeout_secs: int = 2):
         self.step_client = self.node.create_client(GymStep, 'gym_step')
         self.reset_client = self.node.create_client(GymReset, 'gym_reset')
         self._step_size = 0.1
+        self.goal_client = self.node.create_client(GymGoal, 'gym_goal')
+        self.sphere_axis0 = self.node.create_publisher(msg_type=Float32, topic="/sphere_axis0/sphere_axis0/target")
+        self.sphere_axis1 = self.node.create_publisher(msg_type=Float32, topic="/sphere_axis1/sphere_axis1/target")
+        self.sphere_axis2 = self.node.create_publisher(msg_type=Float32, topic="/sphere_axis2/sphere_axis2/target")
+
+    def _log_robot_state(self, robot_state):
+        q_pos = robot_state.q
+        q_vel = robot_state.qdot
+        qpos_str = str(q_pos).strip('[]')
+        qvel_str = str(q_vel).strip('[]')
+        self.node.get_logger().info("joint angles: %s" % qpos_str)
+        self.node.get_logger().info("joint velocity: %s" % qvel_str)
+
+    def _check_service(self, srv):
```

Comment:
my understanding is that this method never terminates if the service is not available. To timeout after some time, see how other methods use `self._wait_until_future_complete_or_timeout(self, future):`

Reply:
I deleted check service and replaced every call with _wait_until_future_complete_or_timeout

Reply:
You mixed the functions. Check service is checking if the service is available. There are 3 services goal, step and reset. Spin_until_convergence is another function which waits until service responds back.